### PR TITLE
Use json.dumps instead of dump for Py2 compatibility

### PR DESCRIPTION
**Pull request by kdheepak**
This change is backwards compatible with Py2 and Py3. `json.dump` apparently has bugs in Py2 when dealing with unicode, whereas `json.dumps` works correctly in both Python2 and Python3. See [StackOverflow](http://stackoverflow.com/a/34574105/5451769) for reference.

### CODE REVIEW

**Comment by wagnerluis1982** [comment]
I think the pull request title is incorrect: it's the inverse of the commit summary.

**Comment by kdheepak** [comment]
Yes, sorry about that. I'll edit the title.

**Comment by wagnerluis1982** [comment]
And I think it's ok, but on very very long texts, a big memory peak may occur.

**Comment by kdheepak** [comment]
That's right. It looks like `json.dump` does write chunks. Perhaps we
shouldn't merge this either, and let it be part of the `python2` branch.

**Comment by sergiocorreia** [comment]
Looks good.

I'm not that worried about a few Mb of overhead but I'll run a few tests nonetheless to see how big the issue is (I also want to see how both json.dump and json.dumps work internally, as there might be a way to call JSONEncoder directly).

**Comment by sergiocorreia** [comment]
It seems that in py2 there is no `sys.stdout.detach()`, which would require a slight code change ([perhaps](http://stackoverflow.com/questions/4374455/how-to-set-sys-stdout-encoding-in-python-3) just removing `.detach()` is enough, but I'm not sure).

I also tried to play a bit with something that worked on both py2 and py3 but got stuck in unicode hell, so given all this my best bet would be to just maintain this part separated and leave the existing py3 code as-is (panflute has all the IO code in a separate file partly because of this concern).

**Comment by sergiocorreia** [comment]
Update: I added some big markdown files and then profiled panflute.

It turns out that most time was spent on `json.dump()`. It uses chunks which meant low memory **but also made it a low slower**.

After testing your PR, the time required to run [a simple filter](https://github.com/sergiocorreia/panflute/blob/master/tests/input/awesome-c/profiler.py) went from 1.6s to 1.0s (an improvement of almost 40%!)

At the same time, memory usage never went above 50mb (for a file with 100k elements), so the memory cost is not that high given the time savings.

**Comment by kdheepak** [comment]
Great! Thanks for being so responsive! Do you think it's worthwhile to document and track your efforts with Py2/Py3 unicode in a separate Github issue?

**Comment by sergiocorreia** [comment]
I kinda gave up on that, as my py2 knowledge is quite rusty. Did you manage to get Unicode strings working on py2? If not, let me know what errors you find and fix can try digging up.

**Comment by kdheepak** [comment]
hi @sergiocorreia, the version in the `python2` branch works now! With Python2 and Python3. I've only run the tests you have in there, and have not actually used it for a project. The commit from this pull request was all that was needed to fix the problem. Are you still having issues? Can you test it with the `python2` branch in there currently?

**Comment by sergiocorreia** [comment]
Have you tried running a filter through a pandoc call? Just want to be sure there is no other differences like the sys.stdout.detach() :)

**Comment by kdheepak** [comment]
Good call! I made some changes but haven't pushed them to the `python2` branch yet. They are on my fork. Some of the examples in `examples/make.py` work with these changes. If I can get all the examples working I'll push them to the `python2` branch.
